Src/Web/Web.Net45/AspNetDiagnosticTelemetryModule.cs

```diff
@@ -0,0 +1,211 @@
+
+        /// <summary>
+        /// Indicates if module initialized successfully.
+        /// </summary>
+        private bool isEnabled = true;
+
+        /// <summary>
+        /// Initializes the telemetry module.
+        /// </summary>
+        /// <param name="configuration">Telemetry configuration to use for initialization.</param>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            try
+            {
+                foreach (var module in TelemetryModules.Instance.Modules)
+                {
+                    if (module is RequestTrackingTelemetryModule)
+                    {
+                        this.requestModule = (RequestTrackingTelemetryModule)module;
+                    }
+                    else if (module is ExceptionTrackingTelemetryModule)
+                    {
+                        this.exceptionModule = (ExceptionTrackingTelemetryModule)module;
+                    }
+                }
+            }
+            catch (Exception exc)
+            {
+                this.isEnabled = false;
+                WebEventSource.Log.WebModuleInitializationExceptionEvent(exc.ToInvariantString());
+            }
+
+            this.allListenerSubscription = DiagnosticListener.AllListeners.Subscribe(this);
+        }
+
+        /// <summary>
+        /// Implements IObserver OnNext callback, subscribes to AspNet DiagnosticSource.
+        /// </summary>
+        /// <param name="value">DiagnosticListener value.</param>
+        public void OnNext(DiagnosticListener value)
+        {
+            if (this.isEnabled && value.Name == AspNetListenerName)
+            {
+                var eventListener = new AspNetEventObserver(this.requestModule, this.exceptionModule);
+                this.aspNetSubscription = value.Subscribe(eventListener, eventListener.IsEnabled);
+            }
+        }
+
+        /// <summary>
+        /// Disposes all subscriptions to DiagnosticSources.
+        /// </summary>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+        
+        #region IObserver
+
+        /// <summary>
+        /// IObserver OnError callback.
+        /// </summary>
+        /// <param name="error">Exception instance.</param>
+        public void OnError(Exception error)
+        {
+        }
+
+        /// <summary>
+        /// IObserver OnCompleted callback.
+        /// </summary>
+        public void OnCompleted()
+        {
+        }
+
+        #endregion
+
+        private void Dispose(bool dispose)
+        {
+            if (dispose)
+            {
+                this.aspNetSubscription?.Dispose();
+                this.allListenerSubscription?.Dispose();
+            }
+        }
+
+        private class AspNetEventObserver : IObserver<KeyValuePair<string, object>>
+        {
+            private const string FirstRequestFlag = "Microsoft.ApplicationInsights.FirstRequestFlag";
```

Comment:
1. RequestTelemetry is set in the BeginRequest and always in the context at the time of EndRequest. So it cannot be used as an indicator of the first request.
2. Why it could happen: 
AppInsights.Web protects itself from the case when module is added twice into the web.config.
It is questionable whether there should be protection from it, but this is already there and there are even functional tests for it. I don't want to break it.
So the AspNetEventObserver just repease the logic of ApplicationInsightsHttpModule.

TelemetryCorrelationModule, being generic module anyone could consume, does not try to protect listener from it.